### ISSUE TITLE
Pyright fails to infer argument type of Callable within a NamedTuple

### ISSUE THREAD

**Issue by MKuranowski** [microsoft/pyright#7887]
**Describe the bug**
Instantiating NamedTuples with a Callable field with a lambda expression causes _Argument type is unknown_ on the lambda expression argument. This doesn't happen with dataclasses.

**Code or Screenshots**

```python
from dataclasses import dataclass
from enum import IntEnum
from typing import Any, Callable, NamedTuple


class Color(IntEnum):
    RED = 0
    GREEN = 1
    BLUE = 2


@dataclass
class DataclassConverter:
    name: str
    fn: Callable[[str], Any]


class NamedTupleConverter(NamedTuple):
    name: str
    fn: Callable[[str], Any]


DataclassConverter("foo", lambda x: Color(int(x)))
NamedTupleConverter("foo", lambda x: Color(int(x)))
```

Results in:

```
$ pyright ignore_test.py 
/.../ignore_test.py
  /.../ignore_test.py:24:35 - error: Type of parameter "x" is unknown (reportUnknownLambdaType)
  /.../ignore_test.py:24:48 - error: Argument type is unknown
    Argument corresponds to parameter "x" in function "__new__" (reportUnknownArgumentType)
2 errors, 0 warnings, 0 informations 
```

This issue is reported only in the NamedTupleConverter. `x` is clearly `str` in this case, and pyright has no trouble inferring that in the `DataclassConverter`.

**VS Code extension or command-line**

1.1.362, installed from PyPI.

**Comment by erictraut** [comment]
Both named tuples and dataclasses have synthesized `__new__` and `__init__` methods. Until recently, pyright synthesized a simple `__new__` signature (`*args: Any, **kwargs: Any`) and a more complete `__init__` method for both. It was pointed out in [this issue](https://github.com/microsoft/pyright/issues/7769) that this didn't match the runtime behavior. Consequently, pyright now synthesizes a simple `__init__` signature and a more complete `__new__` method for named tuples.

When both a `__new__` and `__init__` method are present, argument types are inferred with bidirectional type inference using the `__init__` signature, not the `__new__` method. This is almost always preferable. However, in this case, it results in the behavior you're seeing.

To work around this, you can disable `reportUnknownArgumentType` and `reportUnknownLambdaType` type checks, suppress these diagnostics using a `# type: ignore` or `# pyright: ignore` comment, or use a temporary variable so the lambda's type can be inferred.

```pyright
conv: Callable[[str], Color] = lambda x: Color(int(x))
NamedTupleConverter("foo", conv)
```
